src/MessageWrapper.php

```diff
@@ -0,0 +1,94 @@
+<?php
+namespace PhoneBurner\Http\Message;
+
+use Psr\Http\Message\MessageInterface;
+use Psr\Http\Message\StreamInterface;
+
+trait MessageWrapper
+{
+    private $message;
+    private $message_factory;
+
+    protected function setFactory(callable $factory)
+    {
+        $this->message_factory = $factory;
+    }
+
+    private function viaFactory(MessageInterface $message)
+    {
+        if (!$this->message_factory) {
+            return $message;
+        }
+
+        return call_user_func($this->message_factory, $message);
+    }
+
+    protected function setMessage(MessageInterface $message)
+    {
+        $this->message = $message;
+    }
+
+    private function getMessage(): MessageInterface
+    {
+        if (!($this->message instanceof MessageInterface)) {
+            throw new \UnexpectedValueException('must `setMessage` before using it');
+        }
+
+        return $this->message;
+    }
+
+    public function getProtocolVersion(): string
+    {
+        return $this->getMessage()->getProtocolVersion();
+    }
+
+    public function withProtocolVersion($version): MessageInterface
```

Comment:
COULD: Would we want to enforce the implicit interface of the `MessageInterface` defined by the docblocks, since the PSR predates scalar type delarations?  That is, this method `withProtocolVersion` has a docblock that declares that `$version` is a string.  Do we want to enforce that intention strictly?
```suggestion
    public function withProtocolVersion(string $version): MessageInterface
```

Also, thoughts on declaring strict types throughout the package?

Reply:
Interesting. I'm :+1: on adding scalar types, I'm not sure what declaring strict types would do - they'd only apply to the _trait_, right? Not the classes that `use` the trait?

Reply:
That is my understanding -- it would only affect the trait provided methods, not the classes consuming the trait.